Validation loss

rockerBOO opened the pull request:
Related #1856 #1858 #1165 #914

Original implementation by @rockerBOO 
Timestep validation implementation @gesen2egee 
Updated implementation for sd3/flux @hinablue

I went through and tried to merge the different PR's together. I **_probably_** messed up some things in the process. 

One thing I wanted to note is that `process_batch` was made to limit duplication of the code for validation and training to keep them consistent. I implemented the timestep processing so it could work for both. Noted that it was using only debiased_estimation in other PR's but i didn't know why it was like that.  

`train_db.py` and `train_controlnet.py` I did not update appropriately to my goal of a unified `process_batch`, as I do not have a good way to test them. I will try to get them in an acceptable state and we can refine it. 

I'm posting this a little early so others can view and give me feedback. I am still working on some issues with the code so let me know before you dive in to fix anything. Open to commits to this PR, can post them to this branch on my fork. 

## Testing
- Test training code is actually training
- Test validation epoch
- Test validation per step
- Test validation split
- Test validation seed
- Test image latent caching
- Test tokenizing strategy
- Test text encoding strategy
- Test `--network_train_text_encoder_only`
- Test `--network_train_unet_only`
- Test training some text encoders (I think this is a feature?)
- Test on SD1.5, SDXL, Flux models
- Test on SD1.5, SDXL, Flux LoRAs

I'm open to feedback about this approach and if anything needs to be fixed in the code to be accurate.

gesen2egee commented:
Hi, thank you for your work!

The primary reason for using debiased_estimation_loss is to rescale the loss at different time steps during SD training, allowing better reflection of high-timestep signals. However, after comparing the results with simple averaging, I found that apart from the magnitude, the linear artifacts didn't show significant differences. Perhaps fixing (t=500) could be sufficient. This could save time and increase the number of samples per run, potentially resulting in smoother outcomes.

The motivation behind designing a separate function for the validation batch is to lock in the hyperparameters, facilitating direct comparisons across different hyperparameter configurations during training.

Additionally, I'm not quite sure about the purpose of modifying train_util's timestep and noise part, as it is a library used by many training scripts. Changing the function names might cause other functionalities to stop working. In fact, I don't really agree with modifying too much shared code, as it raises the barrier to merging with other PRs.

rockerBOO commented:
For the loss calculations I think keeping it consistent to the training even if it may be more or less impactful at different timesteps. I feel this could be iterated later, if a separation is necessary.

The process_batch was to unify the training and validation to keep them consistent in their process. This prevents having to keep 2 places updated for batch processing for both. Note https://github.com/kohya-ss/sd-scripts/pull/914#issuecomment-1793725054 where this was discussed. 

For the inconsistencies in train_util, there was the change of `get_timesteps` and `get_huber_threshold_if_needed`. 

`get_timesteps` did not exist previously so my implementation of `get_random_timesteps` which is similar was to implement this behavior. We can use the original or rename it to be more aligned to the args component providing context to different timesteps behavior. Also we convert the timesteps into long() but the implementations want a IntTensor, not sure why that is like that besides maybe historically set like that. Probably not a big deal ultimately but not having to convert it might be simpler.

`get_huber_threshold_if_needed` also didn't exist previously, which is similar to the created `get_huber_c` I created. I think we can use the upstream version and remove the `get_huber_c`. 

I added more abstractions of different features of the larger functions like `get_noise_noisy_latents_and_timesteps` to do the individual components. This allows us to utilize them individually like I have translated into process_batch for timesteps. This allows us to keep consistent behavior and limit having similar behavior duplicated and having to be updated together. 

At this time, the following code I want to revert in `process_batch`, which should be a little cleaner. 

This code extracts a few things, including timesteps, which makes it hard to apply our own timestamps here. I implemented all the components in `process_batch` but maybe some compromise in the approach here would be good. 
```python
noise_pred, target, timesteps, weighting = self.get_noise_pred_and_target(
                        args,
                        accelerator,
                        noise_scheduler,
                        latents,
                        batch,
                        text_encoder_conds,
                        unet,
                        network,
                        weight_dtype,
                        train_unet,
                    )
```

Utilizing this function for the various loss adjustment parts. 

```python
loss = self.post_process_loss(loss, args, timesteps, noise_scheduler)
```

In terms of modifying the code, if you could reference to any PR's you may know of that would conflict with these changes (after reverting/fixing the 2 functions `get_timesteps` and `get_huber_threshold_if_needed`), I can help to update or align the changes in this PR to be better. Should mostly be new additions and type signatures on some of the functions. 

Let me know what you think of these.

rockerBOO commented:
I'm working through some bugs with the process but one additional concern is how wandb does it's step calculations. If you provide a step that is not sequential (after the current stated step) it will not record that information at all. It states a warning but nothing gets logged. 

![2025-01-03_15-08](https://github.com/user-attachments/assets/1d0124dd-2aed-4683-a0e2-1bdf0ac46475)


For example if we are stepping through
```
accelerator.log({'epoch_loss': 0.1}, step=global_step)
```

And then we want to set the step for the log in an epoch, it will not record this.

```
accelerator.log({'epoch_loss': 0.1}, step=epoch)
```

The "fix" is to create wandb metrics. https://docs.wandb.ai/support/log_metrics_two_different_time_scales_example_log_training/ . This would allow the different metrics but for other logging like tensorboard it wouldn't work right. Setting the step= to anything besides a higher value than the last recorded loss will fail. 

If we remove the step value we can be a little more flexible by allowing it to set a larger step each time it logs. But this makes the graphs pretty non-usable... 
![Screenshot 2025-01-03 at 15-04-58 apricot-terrain-143 landscape-kohya-lora – Weights   Biases](https://github.com/user-attachments/assets/7572bc39-1dae-45e5-b1e3-2783d9f8bfed)

Ultimately I am not sure what will solve all the options without testing each tracker to make sure it is working as well as update all the tracking `accelerator.log` to accommodate all these. Without updating these wandb won't get validation results, epoch, or validation epoch loss graphs.

rockerBOO commented:
At this point train_network.py should be in parity will sd3 upstream and the associated validation/training through process_batch. 

I reverted/refactored the names of functions in train_util to keep them as they were in the current sd3 upstream. There are added functions that decoupled from the larger noise_noise_latents_timesteps function. 

I set all accelerator.log() in train_network to drop the steps but left the original there for testing. See the comment above about accelerator.log issues. These need to be fixed before release.

gesen2egee commented:
I'm not entirely sure if it's necessary for the time steps of the training and validation sets to be consistent, since they should be independent, right?
On the contrary, I'm more concerned about the loss variations caused by different batch samples being randomly assigned to different time steps, as this can be significant at different SNR levels.
![image](https://github.com/user-attachments/assets/93f46f80-d54c-44c7-98e6-f27f5c33b810)

That's why I previously used fixed, averaged, and debiased methods.

By the way, I tested it yesterday with t=500, and it seems to work?
![image](https://github.com/user-attachments/assets/f908fb98-cb82-4b5a-9c5b-c4de374e97e6)

rockerBOO commented:
The timesteps are random for training. 

The validation i pass a `timesteps_list` which makes them fixed. https://github.com/kohya-ss/sd-scripts/pull/1864/files#diff-62cf7de156b588b9acd7af26941d7bb189368221946c8b5e63f69df5cda56f39R378-R388 

I have them averaging over the timesteps presented.https://github.com/kohya-ss/sd-scripts/pull/1864/files#diff-62cf7de156b588b9acd7af26941d7bb189368221946c8b5e63f69df5cda56f39R457-R459

 Regular training will produce one random timestep. This is averaged the same to produce the expected result. I do not think debiased estimation is necessary to be specific for validation. Whatever the user is training with will be used in validation which should keep them consistent. Since it's doing a range of timesteps it should average out in line with their chosen post process loss function. Is that the confusion? Or is there something else I'm missing?

The loss variations seems to be appropriate for different timesteps? The batch should be processed together for each of the fixed timesteps and averaged out. It seems to produce an appropriate result in the testing I have done so far. 

What are you refering to as t=500? The fixed timesteps for validation to be only 500? Also I think it might be good for fixed timesteps to be configurable with a default so the user can set it or they can set the number of timesteps to test and we can distribute the timesteps across the whole 1-1000 range. 

Let me know what you think. Thanks for the charts.

rockerBOO commented:
@67372a Thanks for sharing these.

As an overall note I think part of the confusion is what is validation support to do, or the definition I am using here. Validation is to test differences between the training dataset and a separate validation dataset in terms of their loss. This is to highlight the differences between the training and validation to highlight overfitting of the training dataset. 

Validation is then the same exact process of the training dataset but we do not do training on it (so no gradient or backwards loss). This gets us a clear distinction with the minimal differences between the runs. 

I think what the confusion is that Validation sounds like it should make sure everything is correct. There is another pass usually that evaluates the results in a very consistent manor, and produces a metric that can be compared between runs. Like an eval or test run that is a separate dataset from training or validation. For generative image AI that is usually FID or CLIP score or some other options that may be available.

Maybe that is the confusion of the goals of validation vs training and validation vs eval/test. That is why validation needs to be consistent to the training because it's suppose to highly the differences between the overfitting of the training dataset and needs to be as consistent as possible to highly that overfitting.  The validation can also "overfit" eventually which can highlight why a third dataset might limit how much the validation dataset is overfitting. At some point, for many purposes of fine tuning, is that the dataset can be very small so we want to be flexible with what people decide to do here. Also eval/test runs can be run using the model inference and comparing prompt/samples, not requiring another dataset or splitting the current dataset. 

I have wanted to add an eval/test run using these metrics to be able to concretely compare different runs which should include a lot of the suggested behavior that was being applied to validation runs. 

- https://developers.google.com/machine-learning/crash-course/overfitting/overfitting

rockerBOO commented:
>     * Always start from the same validation seed when a validation sequence runs, preserving rng states before, and reapplying after. The logic to allow randomization of the dataloader seed based on global step is not required, it was implemented to allow a niche use case (caption modification, shuffling, dropout, etc) that I allowed for, but don't recommend.
>       https://github.com/67372a/sd-scripts/blob/ab6a33ac6de50efbdcc347644d0ce27f83201e44/train_network.py#L605
>       Why: To make val loss more consistent, otherwise RNG can cause it to vary more, AND to avoid impacting non-val rng

The validation seed has been used for shuffling the dataset. It's a minor change but can keep the validation dataset consistent if your training run has different seeds. This is important because the seed is used in random latent generation, and other random factors which you may modify for other reasons. 

If may be a good idea to apply the validation seed for the dataset specific random options as you suggested, caption dropout, caption shuffling and so on. 

>     * Always make sure to flip the network to eval before running val loss, and train after. You can do that in the function if you like, I did it outside to cover samples as well.
>       https://github.com/67372a/sd-scripts/blob/ab6a33ac6de50efbdcc347644d0ce27f83201e44/train_network.py#L1600
>       Why: If the network has dropout and it is set to train, dropout will be applied, which is not desirable.

The model is set to not have a gradient but because I unified the training/validation to be processed in a consistent manner it is a little more nuanced in how it is approached. The torch.set_grad_enabled() allows us to toggle it based on the conditions of training or validation, and if we are training specific parts of the model. 

Dropout is applied in the back propagation so wouldn't be a factor for forward inference done in validation.
 
>     * Set to statically use L2 loss and NOT apply any loss modification during process_val_batch (e.x. min snr gamma, debiased, etc), the exception being that v-pred and ztnsr should still apply if enabled. I also decided not to apply loss masking, but I could see that going either way. I also disable any noise modifications (offset, multires, ip noise gamma, etc) for the val sequence, which I did by going through the function calls and adding a train parameter, which is true by default.
>       https://github.com/67372a/sd-scripts/blob/ab6a33ac6de50efbdcc347644d0ce27f83201e44/library/train_util.py#L6259
>       Why: To be able to compare val loss across training loss modification settings, different loss types, and noise modification settings, we need val loss to remain unmodified. v-pred and ztsnr are the exception, at least from my testing, as they fundamentally change the training dynamic, so if they aren't set, val loss seems to break.

The loss modifications are to remain consistent for comparing tests and validation and is very important to remain consistent to highlight overfitting of the training dataset. 

>     * Use torch inference mode context, and make sure anything called is not later enabling grads (using train parameter as mentioned before).
>       https://github.com/67372a/sd-scripts/blob/ab6a33ac6de50efbdcc347644d0ce27f83201e44/train_network.py#L477C9-L477C65
>       Why: We don't want to generate gradients or apply autograd, this disables both, and may provide a performance benefit. See https://pytorch.org/docs/stable/notes/autograd.html#locally-disable-grad-doc

As mentioned in the current implementation the gradients are turned off for validation. Note that train_network.py has my intended implementation but train_db I haven't updated to make sure we are all on the same page of the implementation process.

rockerBOO commented:
Ahh that seems accurate about dropout in forward and backward. I thought it was backward only.

I still think it's relevant to have it be the same value though because it is specifically aligned with overfitting, so your validation results would be different if you didn't have it be the same amount as the training dataset.

rockerBOO commented:
One thing I'm noting here to resolve is that `self.get_noise_pred_and_target()` also gets the timesteps. If timesteps is decoupled, this would allow us to utilize this abstraction while independently creating the timesteps where appropriate. This would allow higher compatibility with sd3 and flux scripts which may have different implementations.

rockerBOO commented:
To allow this to be completed and support SD3 and Flux I decided to drop fixed_timesteps for validation. It adds a bunch of refactoring to support appropriately, but can be updated to support this after. This tries to limit the updated/refactored code so we can get this released, and we can iterate on it after. 

I also reverted train_db.py mostly to limit the changes in this PR. If the process_batch change is accepted, then each training script will need to be updated, and we can iterate on that.

rockerBOO commented:
Added 
`--validate_every_n_epochs` 
Changed `--validation_every_n_step` to `--validate_every_n_steps`

rockerBOO commented:
Last bit is the `accelerator.log()` factor but otherwise should be in a good state now

rockerBOO commented:
I tried to make the accelerate.log() work approrpiately for wandb and tensorboard. Since wandb will drop logging if the step value is not the same or increasing value, it feel this compromise is OK. 

Added the following wandb metrics for epoch and val_step. This allows us to check the values of those in correlation and update the charts to reflect this. 

```python
            wandb_tracker = accelerator.get_tracker("wandb", unwrap=True)

            # Define specific metrics to handle validation and epochs "steps"
            wandb_tracker.define_metric("epoch", hidden=True)
            wandb_tracker.define_metric("val_step", hidden=True)
```

See how this shows "epoch" on the bottom vs the normal "steps".

![Screenshot 2025-01-07 at 22-32-57 astral-bee-160 landscape-kohya-lora – Weights   Biases](https://github.com/user-attachments/assets/686dbc45-1106-4c3b-b4d2-e1db2a1b21ce)

Which allows us to be able to set epochs to the X axis, which aligns it as expected. 

![Screen Shot 2025-01-07 at 22 37 01](https://github.com/user-attachments/assets/c7e2c3d1-edb8-4165-aeb7-d67a5bb11822)

Only issue right now is for "val_step" incrementing because we do multiple runs on the same "global_step" for this validation current metric it is only recording the last one here. Probably something to be fixed but this value has a lot of noise anyways, and it's the only one that effected right now. 

![Screenshot 2025-01-07 at 22-39-06 astral-bee-160 landscape-kohya-lora – Weights   Biases](https://github.com/user-attachments/assets/6d7111ee-f48a-4fdd-bd3b-47805442a73d)

On tensorboard it might require more work to make it work as well as wandb but I think it's ok for now.

![Screenshot 2025-01-07 at 22-41-01 TensorBoard](https://github.com/user-attachments/assets/668a8efa-3bf4-47bb-a06f-9fe697d8295d)

Before we would set the "epoch" as the step value `accelerator.log(log, step=epoch + 1)` but this would break wandb. Right now we use the `global_step` value as the step `accelerator.log(log, step=global_step)` and in `log = {"epoch": epoch + 1}` but this isn't causing that value to be reflected in the similar spot as it was previously. 

Other factor is it adds metrics for `epoch` and `val_step` in Tensorboard, but can be minimized to allow other trackers like wandb the flexibility. Maybe later it would be appropriate to have this be more nuanced than through `accelerate.log` directly. 

At this point everything should be completed and ready for a full review.

rockerBOO commented on train_network.py:
```diff
@@ -114,15 +117,18 @@ def generate_step_logs(
                     )
 
         return logs
 
     def assert_extra_args(self, args, train_dataset_group):
-        train_dataset_group.verify_bucket_reso_steps(64)
+        # train_dataset_group.verify_bucket_reso_steps(64)
+        # TODO: Number of bucket reso steps may differ for each model, so a static number won't work
+        # and prevents models like SD1.5 with 64
+        pass
```
I have been using bucket_resp_steps of 32 in my dataset previously, but this was set to 64 so it broke it. I was refactoring some code at the time so I added this TODO but if its for SD1.5 I think 32 would be appropriate here?

rockerBOO replied:
From the diffusers library https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L261-L281 Which seems to highlight that 64 is correct and 32 is not correct. I will update the code to be 64 as it was.

rockerBOO commented:
> It seems that SD1.5, SDXL, SD3, FLUX.1, and ControlNet etc. training no longer work with this PR. Are there any plans to support them?

I'm not sure what you mean in that the training no longer works? I have been testing with SD1.5 but would expect all these to work. The dataset issue you mentioned above could be causing an issue with getting all the datasets. But I would expect that they all would work at least. If you can hint at what is not working I can try to resolve it. 

I have added your suggestions in the review. 

I set `train_dataset_group.verify_bucket_reso_steps(32)` for the base train_network.py to support previous SD1.5 behavior. If you feel 64 is more appropriate I can update my older datasets that had it at 32. 

Thanks for the review, apologies for the mistakes.

rockerBOO commented:
I have updated all calls to `config_util.generate_dataset_group_by_blueprint` to handle extracting from the Tuple and added the return type to that function to help with typechecking. Thanks for pointing it out.

rockerBOO commented:
![Screenshot 2025-01-12 at 20-11-19 cyberpunk-boo-kohya-lora Workspace – Weights   Biases](https://github.com/user-attachments/assets/65c1b7f4-9eb2-40e3-8477-04237aa1cd93)

Added divergence value for step and epoch, indicating the difference between training and validation. Will make it easier to see the difference and not have to rely on overlapping. Maybe a different term would be better as divergence might indicate how much it's moving apart and away from convergence. Also might be better to invert the current to go to the other way to match the loss values. 

Fixed a bunch of things with regularization images datasets and repeats. Fix some issues with validate every n steps (which is important when using repeats and regularization images)

rockerBOO commented:
Bug: If text encoders are cached and validation is enabled, and validation is running the process_batch

 `input_ids = [ids.to(accelerator.device) for ids in batch["input_ids_list"]]` 

Errors that the batch['input_ids_list'] is None

rockerBOO commented on train_network.py:
```diff
@@ -1302,19 +1367,144 @@ def remove_model(old_ckpt_name):
+
+                if is_tracking:
+                    avr_loss: float = val_epoch_loss_recorder.moving_average
+                    loss_validation_divergence = val_step_loss_recorder.moving_average - avr_loss 
```
I've updated LossRecorder to handle 0 steps to account for this. If no steps have happened it will return 0.

rockerBOO commented:
- Added 0 check for LossRecorder
- Added train_text_encoder/train_unet values for validation batches. 
- Added val_dataset_group to assert_extra_args
- Added numpy<=2.0 to requirements for compatibility (some libraries were not constraining this and caused things to break)

stepfunction83 commented:
Shit, I should've looked here before going nuts on my own: https://github.com/kohya-ss/sd-scripts/pull/1898

rockerBOO commented:
There was some code in previous versions to allow a distribution of timesteps to be set (instead of random) which I think is aligned to what you're suggesting. It could be a single timestep or we can cache a random timestep and use the same one. This was removed to allow us to merge this PR and we can approach it with a new PR to add that as it would be involved with some different systems.

I think having more static options or stable options might be a good idea for limited datasets because the variability of a limit dataset might skew the results. A validation dataset may have a few times like 2-10 so having it pick a couple poor timesteps could cause issues the loss calculation to not highlight the right things. The current idea is to take a distribution of timesteps and average them. Like 50, 250, 500, 600, 900 timesteps. This could also be applied to regular training to smooth out the variance from random timesteps, but with a significantly increased training time. 

Stabilizing the noise of the initial latents may be worth experimenting with but storing those latents could cause the training take up more memory that scales up with validation size. 

Also need to consider how limited the datasets were working with, training time, and goals of what validation is doing for us in training. As the datasets get larger these variance become less particular (more examples) and smoothing out the loss for the charts might be enough to highlight it. 

For loss we also do an average of the other steps in an epoch which smooths out the "current" loss and the same for validation but because of the limited datasets I think having more samples from timesteps can help smooth it out with less data.

stepfunction83 commented:
Why not just reset the seed for each validation run so that the same noise
and timestamps can be used?

On Sat, Jan 25, 2025, 9:50 PM Dave Lage ***@***.***> wrote:

> There was some code in previous versions to allow a distribution of
> timesteps to be set (instead of random) which I think is aligned to what
> you're suggesting. It could be a single timestep or we can cache a random
> timestep and use the same one. This was removed to allow us to merge this
> PR and we can approach it with a new PR to add that as it would be involved
> with some different systems.
>
> I think having more static options or stable options might be a good idea
> for limited datasets because the variability of a limit dataset might skew
> the results. A validation dataset may have a few times like 2-10 so having
> it pick a couple poor timesteps could cause issues the loss calculation to
> not highlight the right things. The current idea is to take a distribution
> of timesteps and average them. Like 50, 250, 500, 600, 900 timesteps. This
> could also be applied to regular training to smooth out the variance from
> random timesteps, but with a significantly increased training time.
>
> Stabilizing the noise of the initial latents may be worth experimenting
> with but storing those latents could cause the training take up more memory
> that scales up with validation size.
>
> Also need to consider how limited the datasets were working with, training
> time, and goals of what validation is doing for us in training. As the
> datasets get larger these variance become less particular (more examples)
> and smoothing out the loss for the charts might be enough to highlight it.
>
> For loss we also do an average of the other steps in an epoch which
> smooths out the "current" loss and the same for validation but because of
> the limited datasets I think having more samples from timesteps can help
> smooth it out with less data.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kohya-ss/sd-scripts/pull/1864#issuecomment-2614185207>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AH2WKO4JS5YFZGFXDRZCBJL2MREOJAVCNFSM6AAAAABUREODZWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMJUGE4DKMRQG4>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

stepfunction83 commented:
Is this only implemented for LoRA training?

feffy380 commented:
> ![Screenshot 2025-01-12 at 20-11-19 cyberpunk-boo-kohya-lora Workspace – Weights Biases](https://private-user-images.githubusercontent.com/15027/402400484-65c1b7f4-9eb2-40e3-8477-04237aa1cd93.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk1NDY0NTksIm5iZiI6MTczOTU0NjE1OSwicGF0aCI6Ii8xNTAyNy80MDI0MDA0ODQtNjVjMWI3ZjQtOWViMi00MGUzLTg0NzctMDQyMzdhYTFjZDkzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE0VDE1MTU1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdlM2MzN2MyZjBlZDhlMjU2MTVmZTE2ODI5ZTVhNjhlYzk2ZDVjNWEyMjcxZjQyOTg0MjE0YTQ5NWJiZDM2NDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.oXwQ7oYI_YXJ55NYY95iAJDWIs5TZXd2OZFstEndE58)
> 
> Added divergence value for step and epoch, indicating the difference between training and validation. Will make it easier to see the difference and not have to rely on overlapping. Maybe a different term would be better as divergence might indicate how much it's moving apart and away from convergence. Also might be better to invert the current to go to the other way to match the loss values.
> 
> Fixed a bunch of things with regularization images datasets and repeats. Fix some issues with validate every n steps (which is important when using repeats and regularization images)

There seems to be a mistake with epoch_divergence. It doesn't match the value recorded for step:
```python
# step_divergence
loss_validation_divergence = val_step_loss_recorder.moving_average - loss_recorder.moving_average

# epoch_divergence
avr_loss: float = val_epoch_loss_recorder.moving_average
loss_validation_divergence = val_step_loss_recorder.moving_average - avr_loss
```

If it's the difference between validation and training loss, shouldn't it be `val_epoch_loss_recorder.moving_average - loss_recorder.moving_average`?
